### ISSUE TITLE
PC speaker emulation doesn't sound the same as a real PC speaker

### ISSUE THREAD

**Issue by AXDOOMER** [chocolate-doom/chocolate-doom#795]
Since we are talking about everyone's relative taste (https://github.com/chocolate-doom/chocolate-doom/issues/793), here's something that may be closer to the scope of the project:

*Actual bug report:*
The emulated PC speaker produces sound that's higher-pitched than real PC speaker sound. Sound coming from good old PC speakers are lower-pitched.

**Comment by fragglet** [comment]
Really? Interesting. It would be good if we had some recordings of Vanilla Doom (under DOS with a real physical speaker) vs Chocolate Doom to see what the difference is.

**Comment by AXDOOMER** [comment]
I'll provide some samples as soon as I have a chance.

**Comment by sneakernets** [comment]
From my NEC Laptop (that isn't full IBM PC compatible, but that's another story) I hear pretty much what Chocolate doom outputs. Unless my perfect pitch isn't so perfect when it comes to PC speaker bleeps. There could be some down-pitching from the acoustics in computer cases, the kind of PC speaker you had (actual speaker vs piezo), and, of course, nostalgia goggles.

**Comment by AXDOOMER** [comment]
I have an old PC with a 2.25 inch magnetic speaker and I believe I have another one with a piezoelectric speaker. I will test them, but what's sure is that PC speaker emulation don't as good to the ears as a real PC speaker does.

**Comment by jmtd** [comment]
Perhaps we need to support putting the speaker beeps out of a separate speaker set, so expert users can mount a set of dedicated PC speaker emulation speakers inside their cases to get the accoustics right (/s)

**Comment by AXDOOMER** [comment]
You can set the sound to "PC speaker" in Chocolate-Setup and run Chocolate-Doom with `sudo` and you'll get sound out of the machine's PC speaker. That's what I do.

**Comment by fragglet** [comment]
Yes, Chocolate Doom can drive a real PC speaker on certain OSes.

One flaw with the emulation code is that we don't do any muffling - the sounds all come out as crisp, ear-piercing square waves. Real PC speakers are usually being heard from inside several levels of plastic.

**Comment by chungy** [comment]
I've usually heard laptops sound either identical or pretty close to Chocolate Doom's emulation. Not very surprising, they're often exposed directly to the exterior of the casing...

I suppose some kind of filter could be applied to muffle the sound, but just like CRTs, this will never sound perfect to everybody's PC. It's just impossible.

**Comment by AXDOOMER** [comment]
I tried but didn't work as expected. The two samples that I recorded using my laptop's microphone sound the same. The real PC speaker was buggy on the computer that I tried so there were a lot of cutoffs. 

http://www.mediafire.com/file/zo9tdmsnzs82rin/pcspk-test.mp3
It's not even worth listening to it. 

To my ear, they don't sound the same (they are pretty close though) and maybe with a higher quality speaker I can record a better audible difference. 

The speaker was an integrated speaker inside a HP t5740 case. The PC speaker and the integrated speaker are the same. 

`HOW TO`
Here's how to enable the beeps on your own Linux PC (most distros disable it because it's annoying): 
Go in `alsamixer` and unmute "Beep". Set the volume to ~60. 

If you have a tower with an integrated speaker, then it's the best way you can test it to see the difference, because the sound comes from the same source, which can be under some plastic, you'd be testing it in the same conditions so you'd get the same acoustics.

**Comment by AXDOOMER** [comment]
Sample 2: http://www.mediafire.com/file/lzcg2eljw6ngz2e/pcspk-test2.mp3

Brief comparison of two sequences: http://www.mediafire.com/file/ue2u67005fww0am/pcspk-demo.mp3
First it's the PC speaker emulation, then it's the real PC speaker. There's a cutoff in the sequence will the pistol fires, but again that's the computer being weird. The difference is a little more audible, but still it's not like you were listening to it for real.

**Comment by AXDOOMER** [comment]
Here a more obvious one: http://www.mediafire.com/file/higvjeu5hv0sazt/pcspk-lat.mp3
I'm alternating between the pistol sound from the emulation and the other one.

**Comment by ryan-sg** [comment]
As someone who played Doom with nothing BUT PC speaker, for at least several months in the 90s, this seems a bit frivolous.  Pull the PC speaker out of the case, and it will sound pretty much identical to the emulated effects.  Plus we all have different sized speakers connected to our sound cards these days, I'll bet the PC speaker sounds from my good ole 3-way floor speakers sounds good and 'rich' compared to a typical $5 pair of discount speakers ;)

The point being though, any attempts to filter and muffle the sound -- on a feature that people likely only turn on for the novelty and then switch back off when their spouse starts throwing painful objects no less -- will never be perfect for everyone's setup.

Perhaps lay your speakers face down on a table, or put them point into corners in a room.....or in a box?

**Comment by AXDOOMER** [comment]
>  it will sound pretty much identical

That's the problem, it's *just* pretty much identical. It could be better. 

> Perhaps lay your speakers face down on a table, or put them point into corners in a room.....or in a box?

LOL. I'll try, but the problem may not be the acoustics.

**Comment by haleyjd** [comment]
An optional filter for this sound backend has been on my mind since it was originally implemented into Eternity. I really cannot use the feature as it is, because after about 10 minutes it causes actual pain in my ear drums. The sharp unnatural waveforms are very hard to listen to. For this reason, I cannot possibly view it as "frivilous," but, I also recognize that emulating the muffling effect of a 1990's computer case is a non-trivial problem. Though, given that a computer case is a solid, some low-pass filtering alone would probably do half the work.

**Comment by sneakernets** [comment]
@AXDOOMER : You're asking for a multitude of things - consider these common options:

[Picture of Speakers](https://upload.wikimedia.org/wikipedia/commons/7/71/PC-Speaker_IMG_8161.JPG)

* IBM PC Speaker, basically a small, shallow cone standard dynamic speaker you'd find in some inexpensive desktop monitors. medium to mid-highs.
* Cone Speaker, used in Compaqs - deep cone, big magnet - serious business, and seriously loud. mid-bass to mid-treble range
* Piezo Speaker - cheap PC Speakers that have the dynamic range of a singing Hallmark card, and just as cheap. Very tinny, no bass whatsoever.

Basically the only way to "emulate" this is to implement a band-pass filter with delay. If we're going with PC cases, a very fast quiet delay would work wonders to emulate the sound bouncing around in the case.

**Comment by ryan-sg** [comment]
> The sharp unnatural waveforms are very hard to listen to. For this reason, I cannot possibly view it as "frivilous," 

I hadn't considered headphone usage when I wrote my original reply, is that what you mean @haleyjd?  Otherwise I certainly never noticed it actually being painful to listen to.

**Comment by haleyjd** [comment]
Nope I get the problem even with spatially separated stereo speakers about 3-4 feet away on my desk turned to a low volume level, given enough time. The unpleasant feeling it creates can last a long time after I'm done playing too. Maybe I'm just "unique," I don't know.

**Comment by AXDOOMER** [comment]
@sneakernets : The IBM PC Speaker is the only thing that I want emulated. The other like the Piezo Speaker, is pure garbage. I had one and I couldn't stand the sound just like the current PC speaker emulation.

**Comment by AXDOOMER** [comment]
Ok, now listen to this. 

https://drive.google.com/open?id=0Bx4S0DIPWMITOVRYWVdlTHVIYUU

You will hear the difference. It is exactly what I hear. The audio from the real PC speaker and from the emulated one both come from the same speaker inside the case of my computer which points at the front of it. The front is made of hexagonal holes (like you see inside hives) and the speaker produces really good sound quality for an integrated case speaker. There is no resonance inside the case or whatsoever, so it's just like real desk speakers.

**Comment by NY00123** [comment]
Hey there,

I think that I've stumbled upon a similar situation beforehand, although the explanation of all cases here may appear to be a bit complicated.

@AXDOOMER Just to be sure, have you always run Chocolate Doom under Linux, using the Linux and SDL drivers for PC Speaker sounds? I assume there's been no run of Vanilla Doom under real DOS / NTVDM / DOSEMU?

Reason I'm asking is, that I think that there's a difference in Chocolate Doom's real PC Speaker outputs when compared to Vanilla Doom, and this is observed in a few of your attached sound outputs.

Also, for DOS apps, I should also note that the output under NTVDM or DOSEMU may be a bit different from real DOS (including e.g., MS-DOS and FreeDOS), and this might matter here (albeit maybe not as much for Doom).

I'll begin with explaining how I believe that transforms between beep values/frequencies are done in a few Softdisk/Apogee/id games, when run under vintage DOS setups:
- Earlier games like Catacomb, DDave (1990) and Keen 1-3: Each new (non-zero) value of a given PC sound effect leads to a new square wave i.e., a separate beep. Even if a value is repeated, you get separate short beeps with the same frequency, rather than one long(er) square wave.
- Less early games, from KDreams and up to Doom: A repeated sound value leads to a single (possibly long) square wave, but when the frequency changes, a new wave begins.

Interestingly, as of today, the above isn't faithfully reproduced in vanilla DOSBox builds (either v0.74 or SVN). This is especially observed in earlier games like Keen 1-3, but also applies to the later ones.
There was even an attempt of mine to fix this (with a few later improvement attempts by ripsaw). Link: http://www.vogons.org/viewtopic.php?t=29603
A different patch by ripa with a better fix can be found here (login required for downloading attachments): http://www.vogons.org/viewtopic.php?p=526898#p526898

Back to Chocolate Doom, I think that the current situation with all drivers is this:
- SDL: Basically behaves in a (mostly) similar manner to vanilla DOSBox, merging successive beeps, even of different frequencies (see the way phase_offset is calculated in pcsound_sdl.c).
- WIN32, Linux, BSD: Beeps are separate, as in Keen 1-3 when run under a vintage DOS setup.

For the SDL and Linux drivers, I have attached a patch that aims to fix this (apparently a GitHub gist is one way to do this, see below). Unfortunately, not only it's less convenient for me to check the WIN32 and BSD drivers, but there's also a great limitation; Basically, a duration parameter is required; You can't just seem to tell the OS a command of this form: "Output a beep of a given frequency until I tell you to stop."

If there's a solution for WIN32 and/or BSD, I probably won't test it.

Link to patch: https://gist.github.com/NY00123/d08b79da44ee12464bb6854c8f7e16d5

**Comment by AXDOOMER** [comment]
The sound that I get from playing Chocolate-Doom on Linux with the true PC Speaker is closer to what I had playing Doom under DOS than the PC speaker emulation. I didn't have time to read the whole stuff that you wrote, but I will check that later. Merry Christmas to everyone reading this.

**Comment by NY00123** [comment]
@AXDOOMER Yeah, I think it should be closer to the real thing, at least in general. Exceptions are possible, one example I have being the DPITEMUP sound. Using Chocolate Doom on Linux with real PC Speaker output, it can be said that it sounds somewhat more "robotic" than Vanilla Doom's output. This is more-or-less how should Keen 1-3 do sounds with vintage DOS setups.

Hope you've all had nice times (if not better)!

**Comment by ryan-sg** [comment]
In terms of pc speaker driver accuracy on BSDs I found this tidbit to be the same on both OpenBSD and NetBSD's manpages, with the same message seemingly expanded in FreeBSD's version (in the BUGS section):

OpenBSD/NetBSD:
```
Due to roundoff in the pitch tables and slop in the tone-generation and timer hardware
(neither of which was designed for precision), neither pitch accuracy nor timings will be
mathematically exact.

There is no volume control.

In play strings which are very long (longer than your system's physical I/O blocks) note
suffixes or numbers may occasionally be parsed incorrectly due to crossing a block
boundary.
```

FreeBSD:
```
     Due to roundoff in the pitch tables and slop in the tone-generation and
     timer hardware (neither of which was designed for precision), neither
     pitch accuracy nor timings will be mathematically exact.  There is no
     volume control.

     The action of two or more sustain dots does not reflect standard musical
     notation, in which each dot adds half the value of the previous dot
     modifier, not half the value of the note as modified.  Thus, a note
     dotted once is held for 3/2 of its undotted value; dotted twice, it is
     held 7/4, and three times would give 15/8.  The multiply-by-3/2
     interpretation, however, is specified in the IBM BASIC manual and has
     been retained for compatibility.

     In play strings which are very long (longer than your system's physical
     I/O blocks) note suffixes or numbers may occasionally be parsed
     incorrectly due to crossing a block boundary.
```

I've only ever tried the pc speaker option via SDL though, rather than the system ones.

**Comment by NY00123** [comment]
@ryan-sg Thanks for your investigation on the BSDs. It does look like merely running chocolate-doom as a superuser isn't sufficient on OpenBSD. Quoting from http://man.openbsd.org/i386_iopl.2:

> i386_iopl() sets the i386 I/O privilege level to the value specified by iopl.
> This call may only be made by the superuser. Additionally, it is only permitted when the securelevel(7) is less than or equal to 0 or the machdep.allowaperture sysctl has been set to a non-zero value. 

Now, admittedly, I basically posted a wall of text earlier, so let's link to the patch again (for the SDL and Linux drivers): https://gist.github.com/NY00123/d08b79da44ee12464bb6854c8f7e16d5

As for BSD and (older) Windows setups, maybe the only fix is doing direct I/O ports as done by the OPL drivers. Might be less reliable, though (and unsupported on newer Windows setups).
